examples/scanvi/scanvi.py

```diff
@@ -0,0 +1,251 @@
+        dims = [num_labels + z2_dim] + hidden_dims + [2 * z1_dim]
+        self.fc = make_fc(dims, use_batchnorm=False)
+
+    def forward(self, z2, y):
+        # this is necessary since Pyro expands y during enumeration
+        z2_y = broadcast_inputs([z2, y])
+        z2_y = torch.cat(z2_y, dim=-1)
+        loc, scale = split_in_half(self.fc(z2_y))
+        scale = softplus(scale)
+        return loc, scale
+
+
+# used in parameterizing q(y | z2)
+class Classifier(nn.Module):
+    def __init__(self, z2_dim, hidden_dims, num_labels):
+        super().__init__()
+        dims = [z2_dim] + hidden_dims + [num_labels]
+        self.fc = make_fc(dims)
+
+    def forward(self, x):
+        logits = self.fc(x)
+        return logits
+
+
+# encompasses the scANVI model and guide as a PyTorch nn.Module
+class SCANVI(nn.Module):
+    def __init__(self, num_genes, num_labels, l_loc=5.0, l_scale=5.0, alpha=0.1, scale_factor=1.0):
+        assert isinstance(num_labels, int) and num_labels > 1
+        self.num_labels = num_labels
+
+        assert isinstance(num_genes, int)
+        self.num_genes = num_genes
+
+        self.latent_dim = 5
+
+        assert isinstance(l_loc, float)
+        self.l_loc = l_loc
+
+        assert isinstance(l_scale, float) and l_scale > 0
+        self.l_scale = l_scale
+
+        assert isinstance(alpha, float) and alpha > 0
+        self.alpha = alpha
+
+        assert isinstance(scale_factor, float) and scale_factor > 0
+        self.scale_factor = scale_factor
+
+        super().__init__()
+
+        self.z2_decoder = Z2Decoder(z1_dim=self.latent_dim, y_dim=self.num_labels,
+                                    z2_dim=self.latent_dim, hidden_dims=[50])
+        self.x_decoder = XDecoder(num_genes=num_genes, hidden_dims=[50], z2_dim=self.latent_dim)
+        self.z2l_encoder = Z2LEncoder(num_genes=num_genes, z2_dim=self.latent_dim, hidden_dims=[50])
+        self.classifier = Classifier(z2_dim=self.latent_dim, hidden_dims=[50], num_labels=num_labels)
+        self.z1_encoder = Z1Encoder(num_labels=num_labels, z1_dim=self.latent_dim,
+                                    z2_dim=self.latent_dim, hidden_dims=[50])
+
+    def model(self, x, y=None):
+        # register various nn.Modules with Pyro
+        pyro.module("scanvi", self)
+
+        theta = pyro.param("inverse_dispersion", x.new_ones(self.num_genes),
+                           constraint=constraints.positive)
+
+        with pyro.plate("batch", len(x)), pyro.poutine.scale(scale=self.scale_factor):
+            z1 = pyro.sample("z1", dist.Normal(0, x.new_ones(self.latent_dim)).to_event(1))
+            y = pyro.sample("y", dist.OneHotCategorical(logits=x.new_zeros(self.num_labels)))
+
+            z2_loc, z2_scale = self.z2_decoder(z1, y)
+            z2 = pyro.sample("z2", dist.Normal(z2_loc, z2_scale).to_event(1))
+
+            l_scale = self.l_scale * x.new_ones(1)
+            l = pyro.sample("l", dist.LogNormal(self.l_loc, l_scale).to_event(1))
```

Comment:
@romain-lopez could you please point me to the logic you use to set the mean/variance of the log-normal distribution for the  cell-specific scaling factor `l`? i want to make sure this is done correctly